services/s3/src/it/java/software/amazon/awssdk/services/SelectObjectContentIntegrationTest.java

```diff
@@ -0,0 +1,143 @@
+        s3.waiter().waitUntilBucketExists(r -> r.bucket(BUCKET_NAME));
+        s3.putObject(r -> r.bucket(BUCKET_NAME).key(KEY), RequestBody.fromString(CSV_CONTENTS));
+
+        s3Async = S3IntegrationTestBase.s3AsyncClientBuilder().build();
+    }
+
+    @AfterAll
+    public static void teardown() {
+        try {
+            deleteBucketAndAllContents(BUCKET_NAME);
+        } finally {
+            s3Async.close();
+        }
+    }
+
+    @Test
+    public void selectObjectContent_onResponseInvokedWithResponse() {
+        TestHandler handler = new TestHandler();
+        runSimpleQuery(handler).join();
+
+        assertThat(handler.response).isNotNull();
+    }
+
+    @Test
+    public void selectObjectContent_recordsEventUnmarshalledCorrectly() {
+        TestHandler handler = new TestHandler();
+        runSimpleQuery(handler).join();
+
+        RecordsEvent recordsEvent = (RecordsEvent) handler.receivedEvents.stream()
+                                                                         .filter(e -> e.sdkEventType() == EventType.RECORDS)
+                                                                         .findFirst()
+                                                                         .orElse(null);
+
+        assertThat(recordsEvent.payload().asUtf8String()).contains("A\nC");
+    }
+
+    private static CompletableFuture<Void> runSimpleQuery(SelectObjectContentResponseHandler handler) {
+        InputSerialization inputSerialization = InputSerialization.builder()
+                                                                  .csv(CSVInput.builder().build())
+                                                                  .compressionType(CompressionType.NONE)
+                                                                  .build();
+
+
+        OutputSerialization outputSerialization = OutputSerialization.builder()
+                                                                     .csv(CSVOutput.builder().build())
+                                                                     .build();
+
+
+        SelectObjectContentRequest select = SelectObjectContentRequest.builder()
+                                                                      .bucket(BUCKET_NAME)
+                                                                      .key(KEY)
+                                                                      .expression(QUERY)
+                                                                      .expressionType(ExpressionType.SQL)
+                                                                      .inputSerialization(inputSerialization)
+                                                                      .outputSerialization(outputSerialization)
+                                                                      .build();
+
+        return s3Async.selectObjectContent(select, handler);
+    }
+
+    private static class TestHandler implements SelectObjectContentResponseHandler {
```

Comment:
Have we given any thought into how we might vend ready-made handlers here? Collecting events into a list seems like a common use case.

Reply:
Yeah I do think there's value in precanned ones like that. I think I had to write this same code in 3 different places :p